### DISCUSSION TITLE
Génération d'une API REST à partir de règles publicodes

### DISCUSSION THREAD

**Discussion by johangirod** [publicodes/publicodes#214] [Ideas]
Edit : il existe maintenant un paquet npm pour exposer des règles et calculs publicodes directement en API

**:point_right:  [Voir le README](https://github.com/betagouv/publicodes/tree/master/packages/api)**

## Limites d'un interpréteur javascript

Aujourd'hui, pour la seule possibilité pour partager des calculs publicodes est de passer par un fichier de règles. Ces dernières peuvent être disponibles via un repo github ou directement dans un paquet (exemple : [`modele-social`](https://www.npmjs.com/package/modele-social) ou [`exoneration-covid`](https://www.npmjs.com/package/exoneration-covid)). 

Dans tous les cas, pour pouvoir exécuter les calculs, il faudra passer par l'interpréteur publicodes, qui est en Javascript. C'est une limitation critique, car seuls les environnements acceptant ce langage pourront donc utiliser les calculs. 

Il est actuellement impossible de : 
- Réutiliser les calculs dans une feuille Excel
- Réutiliser les calculs dans une application autre que javascript (Ruby, Python, Go, PHP, et tous les autres).

## Solution : fournir les calculs sous forme d'API REST

Nous allons travailler à développer un module qui permette d'effectuer des calculs sur une base de règle publicodes en utilisant une API REST. 

Cette API serait fournie sous forme de middleware `express` et exposerait des points d'entrée analogues à l'interpréteur JS publicodes.

Voici la forme envisagée : 

- GET `/rules` : liste l'ensemble des règles disponibles dans la base de règle (avec fonction de recherche ?)
- GET `/rules/{id}` : les informations sur une règle telle que l'on peut les obtenir sur une page de doc (description, titre, note, variables utilisées, règles enfants)
- POST `/evaluate` : qui prend comme arguments la situation et une expression publicodes à évaluer. Cette dernière peut être sous forme de liste.

Qu'en pensez-vous ?

**Comment by mquandalle** [comment]
Il me semble que le versionnage doit apparaître dans l'URL des ressources, donc  `GET /v1/rules`

**Reply by johangirod** [reply]
Oui tout à fait. En revanche, je ne pense pas que ce soit au module publicodes API générique de gérer le versionning. C'est au créateur de l'API de savoir si une modification sur une base de règle doit entraîner un changement de version, et de définir quelle stratégie de maintenance des anciennes versions il souhaite avoir. 

Ainsi il suffira au mainteneur de monter l'api des règles v1 sur le point d'entrée v1, et celle des règles v2 sur le point d'entrée v2.

**Comment by mquandalle** [comment]
Pour `POST /evaluate` une possibilité est de mettre la situation en entrée dans le corps de la requête sous forme de JSON, et les objectifs dans un paramètre d'URL

```
POST /v1/evaluate?goals[]=salari%C3%A9%20.%20r%C3%A9mun%C3%A9ration%20.%20brute
```

**Comment by laem** [comment]
Est-ce que cette discussion n'est pas résolue :) ?

**Reply by johangirod** [reply]
Et bien si :tada: :rocket: !
